File: 0992-subarrays-with-k-different-integers/0992-subarrays-with-k-different-integers.cpp
class Solution {
public:
    int subarraysWithKDistinct(vector<int>& nums, int k) {
        map<int, int> mp;
        map<int, int> mp2;

        int left1 = 0;
        int left2 = 0;

        int right = 0;
        int ans = 0;
        int lastUnique = 0;

        while(right < nums.size())
        {
            mp[nums[right]]++;
            mp2[nums[right]]++;

            while(mp.size() > k)
            {
                mp[nums[left1]]--;

                if(mp[nums[left1]]==0)
                {
                    mp.erase(nums[left1]);
                }
                left1++;
            }

            while(mp.size() >= k)
            {
                mp2[nums[left2]]--;

                if(mp2[nums[left2]]==0)
                {
                    mp2.erase(nums[left2]);
                }

                if(mp2.size() < k)
                {
                    mp2[nums[left2]] = 1;
                    lastUnique = left2;
                    break;
                }

                left2++;
            }

            if(mp.size() == k)
            {
                ans += lastUnique - left1 + 1;
            }

            right++;
        }
        
        return ans;
    }
};

File: 0992-subarrays-with-k-different-integers/0992-subarrays-with-k-different-integers.py
class Solution:
    def subarraysWithKDistinct(self, nums: list[int], k: int) -> int:
        
        def atMost(k):
            hashmap = {}
            count = 0
        
            i,j = 0,0

            for j in range(len(nums)):
                hashmap[nums[j]] = hashmap.get(nums[j],0) + 1

                # while it is invalid
                while len(hashmap) > k:
                    hashmap[nums[i]] -= 1

                    if hashmap[nums[i]] == 0:
                        del hashmap[nums[i]] 
                    
                    i += 1
                
                # How many valid subarrays END at j?
                count += j - i + 1

            return count
        return atMost(k) - atMost(k-1)